### PR TITLE
fix(dot/peerset): fix sending on closed channel race condition when dropping peer

### PR DESCRIPTION
**Pull request by timwu20**
## Changes

- sets channel to nil on close
- nil check before producing to channel

## Tests

```sh
go test ./dot/network
```

## Issues

#2567

## Primary Reviewer

@kishansagathiya

### CODE REVIEW

File: dot/peerset/peerset.go [hunk 1]
```diff
@@ -690,10 +690,12 @@ func (ps *PeerSet) disconnect(setIdx int, reason DropReason, peers ...peer.ID) e
...
-		ps.resultMsgCh <- Message{
-			Status: Drop,
-			setID:  uint64(setIdx),
-			PeerID: pid,
+		if ps.resultMsgCh != nil {
+			ps.resultMsgCh <- Message{
+				Status: Drop,
+				setID:  uint64(setIdx),
+				PeerID: pid,
+			}
```
Comment (qdm12) [inline dot/peerset/peerset.go:698]: Shouldn't the producer close the channel? So somewhere here :thinking:

**Review by kishansagathiya — COMMENTED** [review]
So, channel `ps.resultMsgCh` is used in two goroutines `ps.listenAction(ctx)` and `ps.periodicallyAllocateSlots(ctx)`.

It looks like problems is happening because when a context is done `ps.periodicallyAllocateSlots(ctx)` processes it and closes the channel, but `ps.listenAction(ctx)` does not get to know about it until next for loop iteration.

@timwu20 It looks like you have fixed the problem for `disconnect` action, but the give problem could still occur in other actions such as `removeReservedPeers`, `removePeer` and `incoming`.

We can make a similar fix in the remaining actions or check if context was done using `ctx.Err()` before using the channel.

**Comment by timwu20** [comment]
> So, channel `ps.resultMsgCh` is used in two goroutines `ps.listenAction(ctx)` and `ps.periodicallyAllocateSlots(ctx)`.
> 
> It looks like problems is happening because when a context is done, `ps.periodicallyAllocateSlots(ctx)` processes it and closes the channel, but `ps.listenAction(ctx)` does not get to know about it until next for loop iteration.
> 
> @timwu20 It looks like you have fixed the problem for `disconnect` action, but the given problem could still occur in other actions such as `removeReservedPeers`, `removePeer` and `incoming`.

Thanks for pointing this out.  I'll add the check for all the writes to the channel.  I was just debugging the panic produced by running the tests.

> We can make a similar fix in the remaining actions or check if context was done using `ctx.Err()` before using the channel. I like the later approach better (I don't have a reason why :)).

I'd prefer not to check `ctx.Err()` since in the future I'd rather remove context from the service architecture entirely.

File: dot/peerset/peerset.go (hunk 1 above)
Comment (timwu20) [reply dot/peerset/peerset.go:698]: I think there are problems with the arch of `peerset.Handler` that we should address at some point.  
- Right now this channel is returned by `Handler.Messages()`, but it's not a true pipeline.  If `handler.Messages()` was called again and was consumed, then messages would be split among all readers.  
- IMO the producer is `peerset.Peerset` so the producer is actually closing the channel.  Just on another goroutine.

**Review by qdm12 — APPROVED** [review]

**Review by kishansagathiya — APPROVED** [review]
Looks good! Just two more comments with easy fix!!

File: dot/peerset/peerset.go [hunk 2]
```diff
@@ -712,15 +712,27 @@ func (ps *PeerSet) start(ctx context.Context, actionQueue chan action) {
...
+	go ps.listenActionAllocSlots(ctx)
 }
 
-func (ps *PeerSet) listenAction(ctx context.Context) {
+func (ps *PeerSet) listenActionAllocSlots(ctx context.Context) {
+	ticker := time.NewTicker(ps.nextPeriodicAllocSlots)
+
+	defer func() {
+		ticker.Stop()
+		close(ps.resultMsgCh)
```
[the remark this answers is not in the document]
Comment (timwu20) [reply dot/peerset/peerset.go:723]: `actionQueue` is closed in `Handler.Stop()`.  Seems like it's passed into `PeerSet.start`.  It's kind of an odd design to be honest, but I'd rather not refactor the `Handler` to `PeerSet` relationship.

**Pull request merged**